modules/autoscale/src/main.py

```diff
@@ -0,0 +1,140 @@
+import googleapiclient.discovery
+from cloudevents.http import CloudEvent
+import functions_framework
+from google.cloud import secretmanager
+from logging import getLogger, basicConfig, INFO, error, warning, info
+from xml.etree import ElementTree as et
+from xml.etree.ElementTree import Element
+from os import getenv
+from panos.panorama import Panorama
+import base64
+import json
+
+class init_parameters:
+    """
+    This Class is used for create data structure with values used in further function.
+    """
+    panorama_address: str = getenv('PANORAMA_ADDRESS')
+    secret_name: str = getenv('SECRET_NAME')
+    project_id: str = getenv('PROJECT_ID')
+
+@functions_framework.cloud_event
+def autoscale_delete_event(cloud_event: CloudEvent) -> None:
+
+    # Extracting the log data from the Pub/Sub message.
+    configure_logger()
+
+    pubsub_message = json.loads(base64.b64decode(cloud_event.data['message']['data']).decode('utf-8'))
+    instance_id = pubsub_message['resource']['labels']['instance_id']
+    project_id = pubsub_message['resource']['labels']['project_id']
+    zone = pubsub_message['resource']['labels']['zone']
+    details = get_instance_details(project_id, zone, instance_id)
+    interfaces = details['networkInterfaces']
+    items = details['metadata']['items']
+    igm_name = find_igm_name(items)
+    primary_ip = get_primary_ip(interfaces)
+    info(f"-> Start delicensing process for instance: {instance_id}, primary IP: {primary_ip}")
+
+    # get credentials from Secret Manager and login to Panorama
+    credentials = get_secret()
+    panorama = Panorama(init_parameters().panorama_address, credentials.get('user'), credentials.get('pass'))
```

Comment:
Most of the customers are using Panorama working in HA mode, so I propose to extend logic in similar way as we have for AWS, where Lambda for delicensing can work with:
- single Panorama
- Panorama working in HA mode (primary-secondary)

Regarding second case - we need to check which Panorama is active and then do delicensing only on active instance. Feel free to reuse code from AWS: https://github.com/PaloAltoNetworks/terraform-aws-vmseries-modules/blob/e1764506312062be626631d0cd770c918903d9da/modules/asg/scripts/lambda.py#L467-L480

Reply:
The delicensing Python code is aligned with the code for the other clouds. Now the code supports Panorama HA